Telemetry support for SPFx Toolkit. Closes #570

Saurabh7019 opened the pull request:
## 🎯 Aim

This PR adds telemetry support for SPFx Toolkit to help us understand how the extension is used and improve its features. 

## 📷 Result

**App Insights**
<img width="1910" height="2561" alt="portal-azure" src="https://github.com/user-attachments/assets/0370c1d5-2ee4-4ea6-b4a3-e229aeeffb1f" />

**CSV report**
<img width="833" height="620" alt="image" src="https://github.com/user-attachments/assets/7331c178-518c-4b90-83ca-91d707395f8a" />

## ✅ What was done

- [] Proper way of handling the AI connection string
- [] AI capabilities such as chat participants and LM tools
- [] Usage of `vscode.open` actions
- [X] Created a centralized telemetry service
- [X] Extension lifecycle tracking - Activation, Deactivation
- [X] Authentication actions tracking - Login/Logout
- [X] Usage of actions from the Actions view
- [X] Usage of actions from the Tasks view
- [X] App management actions including adding app catalog, managing tenant-wide extensions, and managing SPFx apps
- [X] Support to disable telemetry through VS Code settings
- [X] Updated documentation

## ✅ How to test
- Set up application insights and get your AI connection string
- Search for key 'aiConnectionString' in package.json and replace the value
- Run the PR locally and click on below actions:
 
<img width="1910" height="3727" alt="vscode-viva-about-telemetry" src="https://github.com/user-attachments/assets/3662cc92-6bee-4171-a3ae-ab9deaf0d901" />

## 🔗 Related issue

Closes: #570

Saurabh7019 committed:
Telemetry support for SPFx Toolkit. Closes #570

Adam-it commented:
uuuu looks interesting.... I will try to give it a check/review next week unless someone is quicker than me🤩

Adam-it commented:
@Saurabh7019 I just did a minor and in this process the dev gets rebased/aligned with main. 
Please perform the rebase on your side before we proceed

Saurabh7019 committed:
address code review feedback

Adam-it commented on src/providers/AuthProvider.ts:
```diff
@@ -52,10 +53,10 @@ export class AuthProvider implements AuthenticationProvider, Disposable {
     );
     subscriptions.push(
-      commands.registerCommand(Commands.logout, AuthProvider.logout)
+      commands.registerCommand(Commands.logout, TelemetryService.withTelemetry(Commands.logout, AuthProvider.logout))
```
TBH i would not track those. 
What would be the benefit of knowing this? 
I think what is more important is what functionalities are used?
Do users use spfx chat participant, or SPFx move app... if not lets remove those

Adam-it reviewed (changes requested):
First of all sorry for the huge hold up 🙏
Second of all, awesome work 👏 I left a few comments we could recheck before we proceed

also:

- Shouldn't we also add [telemetry.json](https://code.visualstudio.com/api/extension-guides/telemetry#telemetry.json) so that our extension will support retrieving a dump of all telemetry using vsce cli
- Did you maybe checked that the extension does not log anything to telemetry if we switch it off in VS Code? 
- Should we consider our own setting for it so that for example someone would like to switch of telemetry only for SPFx Toolkit but leave it on for other extensions?
- Internally, lets open a discussion (best on the related topic on Basecamp) to organize a tenant for SPFx Toolkit with App Insights and how we will maintain it and what costs limits we will set

Adam-it commented:
We just had a new minor release. We should rebase this branch before we proceed

Adam-it committed:
Updates scaffolding form. Closes #667 (#694)

## 🎯 Aim

The scaffolding form is not adapted for all versions of SPFx, and most
probably it will take a lot of time until we do that. For now the aim is
to adapt the scaffolding options in order to inform the user that they
only support latest version of SPFx

## 📷 Result

<img width="1536" height="821" alt="image"
src="https://github.com/user-attachments/assets/79322abb-dc38-40e8-9039-d74db80a6994"
/>


## ✅ What was done

- [X] Added notification to the scaffolding form
- [X] Removed gulp fast serve
- [X] Updated chat participant prompts

## 🔗 Related issue

Closes #667

---------

Co-authored-by: Saurabh Tripathi <saurabh7019@gmail.com>

Adam-it committed:
Refactors list related CLI commands to base on CSV output. Closes #701 (#718)

## 🎯 Aim

The aim of this PR is to update all `list` related CLI commands used by
the LLM to base on CSV command output mode instead of JSON which is more
optimal for LLM and reducses the amount of used context window.

## 📷 Result

<img width="1536" height="817" alt="image"
src="https://github.com/user-attachments/assets/b94bfd13-b8f4-4f6b-936d-25736e46d1e6"
/>


## ✅ What was done

- [X] updates output mode to CSV 

## 🔗 Related issue

Closes: #701

---------

Co-authored-by: Saurabh Tripathi <saurabh7019@gmail.com>

Adam-it committed:
Refactors or removes all usage of yo-rc.json. Closes #711 (#717)

## 🎯 Aim

The aim of this PR clean up all left usage of yo-rc.json file so that
SPFx Toolkit is ready for newer versions of SPFx

## 📷 Result

<img width="1536" height="820" alt="image"
src="https://github.com/user-attachments/assets/7b884b3b-0f2c-43b0-866e-b1ec1b478e24"
/>


## ✅ What was done

- [X] Removes ACE checker
- [X] Refactors using yo-rc.json for getting solution name in ci-cd
action to get the name from package.json

## 🔗 Related issue

Closes: #711

Saurabh7019 committed:
pre-release 4.17.1

Adam-it committed:
Refctors `/setup` chat command to SPFx setup LM tool. Closes #706 (#716)

## 🎯 Aim

The aim of this PR is to refactor the old `/setup` chat participant
command to a dedicated Language Model Tool that will validate and
install all needed dependencies for any version of SharePoint Framework

## 📷 Result

<img width="1536" height="816" alt="image"
src="https://github.com/user-attachments/assets/8d1205a2-0670-4690-ba9b-6d364a655d13"
/>

<img width="485" height="509" alt="image"
src="https://github.com/user-attachments/assets/6ec798cb-4d31-4aa3-8f21-249d34757021"
/>

<img width="505" height="454" alt="image"
src="https://github.com/user-attachments/assets/82341651-ce54-4cc3-b780-d1e41c31452b"
/>

<img width="492" height="636" alt="image"
src="https://github.com/user-attachments/assets/ea7152bd-ea80-4051-8bd9-cd390dc591a2"
/>

<img width="504" height="438" alt="image"
src="https://github.com/user-attachments/assets/0a808371-4ba7-40b5-8b4d-067fb7045d2d"
/>

<img width="501" height="652" alt="image"
src="https://github.com/user-attachments/assets/1c802146-9e2e-425c-b406-4b23a4c27501"
/>


## ✅ What was done

- [X] Removed the `/setup` GitHub Copilot Chat command 
- [X] Added SPFx setup environment Language Model Tool

## 🔗 Related issue

Closes: #706

---------

Co-authored-by: Saurabh Tripathi <saurabh7019@gmail.com>

Saurabh7019 committed:
pre-release 4.17.2

Saurabh7019 committed:
Fixes missing ACE samples issue. Closes #723 (#724)

## 🎯 Aim

This PR fixes issues in `prepare-sample-data.ps1` script to ensure all
ACE samples are included in the `sp-dev-fx-samples.json`.

## ✅ What was done

- [X] Better path handling for `.yo-rc.json`
- [X] Appends results to `$samples` instead of overwriting `$output`

## 🔗 Related issue

Closes: #723

Saurabh7019 committed:
Fixes nvm use error on Windows without version argument. Closes #519 (#691)

## 🎯 Aim

This PR fixes the `nvm use` error on Windows when creating a new SPFx
project via the sample gallery. On Windows, `nvm-windows` does not
automatically read the `.nvmrc`, so the command fails without a version
argument. The fix parses the `.nvmrc` file and sends `nvm use <version>`
with the version number for `nvm` users.. for `nvs` users it checks for
both `.nvmrc` and `.node-version` files and sends `nvs use` which works
on all platforms.

## 📷 Result

NA

## ✅ What was done

- [X] modified `createTerminal` method in `TerminalCommandExecuter.ts`
as agreed in the issue

## 🔗 Related issue

Closes: #519

nicodecleyre committed:
pre-release 4.17.3 (#730)

pre-release 4.16.1

nirav-raval committed:
Removes `/info` command and update documentation. Closes #708 (#726)

## 🎯 Aim

The aim is to totally remove all related code regarding the /info SPFx
Chat Participant command and, in the documentation, replace it with
guidance recommending the use of the CLI for the Microsoft 365 MCP
server instead.

## 📷 Result

<img width="1380" height="902" alt="image"
src="https://github.com/user-attachments/assets/41335670-ee0c-41a2-99fd-bc0dad5f512c"
/>
 
## ✅ What was done

- [X] Removed all code related to the /info SPFx Chat Participant
command
- [X] Cleaned up any references, registrations, and configuration linked
to the /info command
- [X] Updated documentation to recommend using the CLI for Microsoft 365
MCP server for tenant info and management

## 🔗 Related issue

Closes: #708

---------

Co-authored-by: Adam-it <adam.wojcik.it@gmail.com>

Adam-it committed:
pre-release 4.17.4

Adam-it committed:
Adds validate local setup for current project action. Closes #518 (#697)

## 🎯 Aim

The aim is to add a new action that will validate user local setup for
the currently open SPFx project.
For example someone opens some old SPFx sample that uses SPFx 1.16 and
would like to recheck if the current version of node and local
dependencies are valid for that project. The aciton uses `spfx doctor`
command to get the report and similar like the setup action will install
the correct version of node and global dependencies if needed based on
the compatibility matrix

## 📷 Result

<img width="1536" height="815" alt="image"
src="https://github.com/user-attachments/assets/22d601c8-8b57-403b-84e5-33bb3cb4aab2"
/>


<img width="1536" height="810" alt="image"
src="https://github.com/user-attachments/assets/fcd31e92-ee89-4cc2-a4e9-abdd631f22c7"
/>


## ✅ What was done

- [X] Added new action
- [ ] Updated docs 👉// TODO

## 🔗 Related issue

Closes: #518

---------

Co-authored-by: Saurabh Tripathi <saurabh7019@gmail.com>

Saurabh7019 committed:
pre-release 4.17.5

Adam-it committed:
Updates docs for language model tools (#739)

## 🎯 Aim

I noticed we are missing some docs about recently added langauge model
tools and this PR is suppose to fill this gap

Adam-it committed:
Updates extension startup. Closes #733 (#743)

## 🎯 Aim

I know the related issue was still not agreed but this one is quite easy
and hits on my nerve .😁
What I usually do is move the SPFx Toolkit Task Pane with heft and npm
actions to the Explorer view so the tasks sit right along my files I
modified. The problem is that this panel does not show up until the
extension gets active so this happens only when we go to the SPFx
Toolkit view. So this means every time I opened a SPFx project I had to
go to the SPFx Toolkit view and then go back to the explorer view just
to see the task panel. This is quite frustrating.
Folks might have customized the look and feel of VS Code, like moving
the login and management views to some other part of VS Code, In this
case they would also need to go to the view to make them active.
That Is why I propose to just check if a project has a
config\package-solution.json file, and if so, we may assume it is a SPFx
project and make the extension active. If it will still not be an SPFx
project, this will be caught by the IsSpfx helper method in
CommandPanels and will just load the SPFx Toolkit with the functionality
when you are not in a SPFx project

## 📷 Result

<img width="684" height="788" alt="image"
src="https://github.com/user-attachments/assets/bc8f7617-93b4-496d-803c-3a970457c254"
/>


## 🔗 Related issue

Closes: #733

Saurabh7019 committed:
pre-release 4.17.6

Adam-it committed:
Updates token handle. Closes #500 (#745)

## 🎯 Aim

The aim is to fix the way we handle MSAL refresh token. Currently, in
the extension, after we sign in, we are more or less logged in 1 hour
and after that the access token expires, and we need to perform sign in
again.
I noticed it was due to a simple bugi and the main origin and fix up was
done and already published in CLI SPFx Toolkit npm package. No the token
should allow to retain sign in from 4-90 days (depending on tenant
settings)

In order to test this PR is best 
1. to uninstall SPFx Toolkit if you have it. 
2. Then use `vsce package` to build a local vscix package and install
SPFx Toolkit from your local package.
3. Next perform sign in
4. Wait for 1-2 hours and restart VS Code. You should still be signed in

## 🔗 Related issue

Closes: #500

Adam-it committed:
pre-release 4.17.7

Saurabh7019 committed:
add telemetry for validateEnvironmentForProject

Adam-it commented:
we had a new minor release again. Maybe lets keep this PR as draft until I setup a tenant for it and make the changes. 
When we find time lets rebase against main
